src/Build/Graph/ProjectGraph.cs

```diff
@@ -303,17 +334,63 @@ public ProjectGraph(IEnumerable<ProjectGraphEntryPoint> entryPoints, ProjectColl
             return targetLists;
         }
 
+        /// <summary>
+        /// A callback used for constructing a <see cref="ProjectInstance"/> for a specific
+        /// <see cref="ProjectGraphEntryPoint"/> instance.
+        /// </summary>
+        /// <param name="projectPath">The path to the project file to parse.</param>
+        /// <param name="projectCollection">The <see cref="ProjectCollection"/> context for parsing.</param>
+        /// <returns>A <see cref="ProjectInstance"/> instance. This value must not be null.</returns>
+        /// <remarks>
+        /// The default version of this delegate used by ProjectGraph simply calls the
+        /// ProjectInstance constructor with information from the parameters. This delegate
+        /// is provided as a hook to allow scenarios like creating a <see cref="Project"/>
+        /// instance before converting it to a ProjectInstance for use by the ProjectGraph.
+        ///
+        /// The returned ProjectInstance will be stored and provided with the ProjectGraph.
+        /// If this callback chooses to generate an immutable ProjectInstance, e.g. by
+        /// using <see cref="Project.CreateProjectInstance()"/> with the flag
+        /// <see cref="ProjectInstanceSettings.Immutable"/>, the resulting ProjectGraph
+        /// nodes might not be buildable.
+        /// </remarks>
+        public delegate ProjectInstance ProjectInstanceFactoryFunc(
+            string projectPath,
+            Dictionary<string, string> globalProperties,
+            string toolsVersion,
+            ProjectCollection projectCollection);
+
+        internal static ProjectInstance DefaultProjectInstanceFactory(
```

Comment:
Nit: Should this be public? In case callers want to do some stuff but still call the default factory.

Reply:
I would say no, nobody should need this.